### ISSUE TITLE
Basis: Indicate visually unpublished

### ISSUE THREAD

**Issue by laryn** [backdrop/backdrop-issues#2269]
It would be good to have an unpublished node differentiated visually somehow when viewing the page (or summary/teaser for that matter since that is technically possible with Views using 'published or admin').

---

Proposed solution: 
![screen-backdrop-pr-unpublished-post-new](https://user-images.githubusercontent.com/14188307/28021242-884011c8-6587-11e7-8745-1dc353cf7249.png)

---
~PR by @laryn: https://github.com/backdrop/backdrop/pull/1815~
~PR by @quicksketch: https://github.com/backdrop/backdrop/pull/1874~
PR by @jenlampton https://github.com/backdrop/backdrop/pull/1936

**Comment by olafgrabienski** [comment]
I agree. There is already the `.unpublished` class in the HTML (at least for the node, didn't check teaser), so only the CSS style is missing. Could you add something, @wesruv ?

**Comment by wesruv** [comment]
Yes! Hadn't thought about it. I'll whip something up.

**Comment by laryn** [comment]
PR for discussion

**Comment by olafgrabienski** [comment]
Setting a low opacity, as this this PR does, works very well for posts in default configuration, because they contain various elements like post information, tags, and a comment section. Seeing all these elements sort of 'greyed out', you get the picture of unpublished. I'm not sure if you 'get the unpublished picture' also on pages which have only body text. See some screenshots for comparison:

![screen-backdrop-pr-unpublished-post](https://cloud.githubusercontent.com/assets/14188307/24121916/41de5e8a-0dba-11e7-8b8b-d9e208b32878.png)

![screen-backdrop-pr-unpublished-page](https://cloud.githubusercontent.com/assets/14188307/24121846/07376056-0dba-11e7-83ed-499758b6fb16.png)

For the page case it might help to set the opacity even lower than 0.4 but I guess then it's hard to read the text on not so good screens.

**Comment by Dinornis** [comment]
Missing PR?

**Comment by klonos** [comment]
Added PR link in the issue summary.

**Comment by Dinornis** [comment]
Thanks @klonos it almost bit me 😄 

@olafgrabienski I wonder if a light grey background with white _DRAFT_ text over it would do, not sure if that can be done using CSS without an image?

**Comment by olafgrabienski** [comment]
@Dinornis I guess it could be achieved by something like

```
.unpublished:before {
  content: "Draft";
}
```

(and then style it accordingly)

On certain sites "Draft" may have other meanings than "Unpublished", so we should use "Unpublished" or "Not published" in my opinion:

```
.unpublished:before {
  content: "Not published";
}
```

I guess it's however not a good idea to follow the approach, as CSS content is not manageable (e.g. translatable) using the Backdrop interface.

**Comment by Dinornis** [comment]
Maybe something like this?
`.unpublished {
  background-color: #fafafa;
  border: 2px dashed #dbdbd9;
  opacity: 0.7;
  padding: 10px;
}
.unpublished h2::after {
  background-color: #ffff99;
  content: " (Not published)";
  font-size: small;
}`

**Comment by olafgrabienski** [comment]
@Dinornis Sorry for my late reply. I'm still concerned about using CSS content because you can't translate it via the Backdrop interface. On the other hand, it's really hard to visualize the unpublished state without using words.

My conclusion at the moment: Not the theme but the Backdrop node functionality should provide the term "Not published" as a (pseudo) field which then can displayed (or not) using the *Manage display* page. What do you think about it?

**Comment by Dinornis** [comment]
I think that sounds really good, if I understand this correctly 😄 
A two part solution, a combination of CSS changing the visual appearance of a unpublished node and an option on the `Manage display` page that would let the user enable/disable the display of  translatable `Not published` text somewhere near the title of the node?

I like that.

**Comment by quicksketch** [comment]
> My conclusion at the moment: Not the theme but the Backdrop node functionality should provide the term "Not published" as a (pseudo) field which then can be displayed (or not) using the Manage display page. What do you think about it?

With the new age of CSS, I think a better option would be to add the translated string "Draft" as an attribute on the `<article>` tag, either as a `title` or using a data- attribute like `data-draft="Draft"`. Then using the trick from https://css-tricks.com/css-content, we can display it with CSS as a background or overlay using

```
[data-draft]:before {
   content: attr(data-draft);
}
```

And you can use CSS to rotate the pseudo-element if desired or do whatever else you want in terms of styling.

**Comment by quicksketch** [comment]
I filed a PR for a proof-of-concept at https://github.com/backdrop/backdrop/pull/1874. I kept the code contained within Basis as it's fairly minimal and highly opinionated (e.g. using the word "Draft" instead of "Unpublished", z-indexing, styling, etc.)

**Comment by quicksketch** [comment]
Here's what it looks like style-wise:

![image](https://cloud.githubusercontent.com/assets/100206/25791272/0e21ee4c-3374-11e7-829b-2f7f4125fd1b.png)

**Comment by Dinornis** [comment]
👍 I like this very much.

**Comment by olafgrabienski** [comment]
@quicksketch Thanks for your suggestion which sounds great! (Definitely have to study this whole data attribute thing, sophisticated!) Came too late to have a look at the PR sandbox website, but I like the concept a lot!

That said, I'm still not sure about the term "Draft", thinking of sites that use a editing workflow with own states (like "Draft", "Review" and so on) which then may overlap with Basis' "Draft aka Not published". On the other hand, "Draft" is much more handy than "Not published". So, I'm indecisive ...

**Comment by quicksketch** [comment]
> That said, I'm still not sure about the term "Draft", thinking of sites that use a editing workflow with own states (like "Draft", "Review" and so on) which then may overlap with Basis' "Draft aka Not published". On the other hand, "Draft" is much more handy than "Not published". So, I'm indecisive ...

Yeah, I specifically wanted to limit this to Basis for that reason, we're not so much making an API here as just dealing with the out-of-box situation. Sites that had more complicated workflows would need to do their own adjustment. Basis would make for a great template: this approach could easily be scaled out to use multiple states with a different data attribute.

**Comment by olafgrabienski** [comment]
Sounds very good to me! (Which milestone should the issue be, btw?)

**Comment by quicksketch** [comment]
I've put this under the next bugfix release (1.6.4).

**Comment by quicksketch** [comment]
A few issues need to be addressed here:

- The positioning of the Draft label on teasers (we don't want it absolutely positioning on the entire page, but on the node wrapper).
- Right now I positioned the "Draft" pseudo-element underneath the text. This has a problem that if there's an image, it covers it all up. Putting it above the text/images makes links/text unselectable/clickable. But then I found [pointer-events](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events) and now everything is solvable. :smile:

**Comment by quicksketch** [comment]
Teaser version (essentially the same):

![image](https://cloud.githubusercontent.com/assets/100206/25977782/d1e9a954-3672-11e7-9317-ace096a54196.png)

Updated the PR with these changes: https://github.com/backdrop/backdrop/pull/1874.

This PR could be commit-worthy at this point. I added missing docs and the styling matches Basis at this point. Let's see if we can get Basis author @wesruv's opinion on this.

**Comment by docwilmot** [comment]
I wonder about accessibility though since the watermark is faint? Is there another indication in the that the node is unpublished (if not, this would be pre-existing so can be another issue).

I would probably think "Unpublished" will be less confusing - there is nothing else in the UI called "draft" and somebody will be hunting for that eventually.

Otherwise, I would RTBC unless @wesruv disagrees.

**Comment by Dinornis** [comment]
I also think that "Unpublished" would be better than "DRAFT".  This way the wording would match the node edit page.

As a secondary indication maybe adding "Unpublished" after the title may be helpful.  I don't know if through CSS like: `.unpublished h2::after { content: " (Not published)"; font-size: small; }` or through code.

**Comment by quicksketch** [comment]
> Is there another indication in the that the node is unpublished (if not, this would be pre-existing so can be another issue).

Core's system.theme.css includes:
```
.unpublished {
  background-color: #fff4f4;
}
```

But IMO that's not really a great indicator. Basis doesn't include that file however, so it doesn't have any styling on unpublished at all right now.

**Comment by olafgrabienski** [comment]
> I wonder about accessibility though since the watermark is faint

Yes, it's quite faint, I'd make it a little darker, maybe around a color-neutral `#bbb` (not a *Basis* color afaik, but maybe less flickering in comparison), leaving the opacity at `0.5`.

> I would probably think "Unpublished" will be less confusing

Absolutely, but sounds also cumbersome. However, using "Unpublished" seems more reasonable... **If** we do so, the font-size should be adapted from `10em` to around `6em`.

Regarding the styling, I'd suggest to also add something like `transform: rotate(-25deg);`. Looks more like a draft in my opinion, and it's easier to override it than to add it.

> As a secondary indication maybe adding "Unpublished" after the title may be helpful.

I guess that's difficult. Not every node has a displayed title. (Apart from that, `.unpublished h2::after` would apply to regular headings within the body text.)

**Comment by olafgrabienski** [comment]
Adding a screenshot with rotated and `#bbb` colored "Unpublished" for comparison:

![screen-backdrop-pr-unpublished-bbb](https://cloud.githubusercontent.com/assets/14188307/26010229/679e35c6-374c-11e7-8b3f-dc0389563b18.png)

With such a long word, we need probably a media query regarding the font-size. On mobile it'll get partly off-screen otherwise.

**Comment by wesruv** [comment]
I couldn't find other PR's but added some code review to this one:
https://github.com/backdrop/backdrop/pull/1874#pullrequestreview-38002837

**Comment by Dinornis** [comment]
I like Olaf's rotated `UNPUBLISHED` watermark version.

**Comment by tomgrandy** [comment]
I second Olaf's watermark and agree that UNPUBLISHED adds clarity to match the node state.

**Comment by jenlampton** [comment]
I'm conflicted between `unpublished` and `draft`. Draft is a new and foreign word for Backdrop, it's not accurate, but it's also more human and friendly than `unpublished`. I could be convinced on either.

I'm not sure about the rotation on the image. What does that look like for teasers that are only a single line long?

**Comment by quicksketch** [comment]
So it sounds like we're generally on the same page here. "Draft" may sound better/simpler but it's not accurate or the "right" word. Let's go ahead with "Unpublished", something is infinitely better than nothing.

Let's review how this looks on small-ish teasers. We could potentially only add the rotation on the full-page version and not on other view modes if needed.

**Comment by olafgrabienski** [comment]
I've had another look at the PR and experimented on it with Chrome's dev tools. My main goal was to see if the quite long term "Unpublished" works also on smaller screens like tablets and smartphones.

I got good results making the font smaller (and darker) for smaller screens and reducing the (position) `top` value. In my opinion, the results look also good on not so long teasers. I added a couple of media queries to cover some common devices. Here's what I changed in comparison to the current PR:

```
.node[data-status]:before {
  color: #777;
  font-size: 2.5em;
  top: 5%;
  transform: rotate(-20deg);
}
@media (min-width: 22.5em) {
  .node[data-status]:before {
    font-size: 3em;
  }
}
@media (min-width: 25.75em) {
  .node[data-status]:before {
    font-size: 3.25em;
  }
}
@media (min-width: 48em) {
  .node[data-status]:before {
    color: #ccc;
    font-size: 4.5em;
  }
}
@media (min-width: 62em) {
  .node[data-status]:before {
    font-size: 6em;
    top: 10%;
  }
}
@media (min-width: 75em) {
  .node[data-status]:before {
    font-size: 7.5em;
  }
}
```

@quicksketch I saw various code suggestions of @wesruv on the PR, feel free to adapt the above suggestions to them.

**Comment by jenlampton** [comment]
I just re-read all the comments. Have we addressed the accessibility issue? By using a watermark of any kind, I wonder if we're always going to have a problem with that. One of my sites that makes heavy use of workflows has added a panel pane with relevant information for each post:

<img width="1198" alt="screen shot 2017-07-06 at 12 14 18 pm" src="https://user-images.githubusercontent.com/397895/27928652-ec78bfaa-6244-11e7-971b-0cafdff57da5.png">

I'm not sure if something like this would be "pretty enough" for Backdrop, but if we like the idea it might be possible to add the "unpublished" text via a ::before pseudo element, and style it to look brightly colored like this.

**Comment by olafgrabienski** [comment]
@jenlampton I like your alternative approach, it's visually more clear. Depending of the styling and visual structure of a post or a page a simple "Unpublished" at the top may however not be clear enough. To make it even more clear, I suggest to add the node type to the `data-status`. Combined with your idea, an unpublished node could for instance like this:

![screen-backdrop-pr-unpublished-post-new](https://user-images.githubusercontent.com/14188307/28021242-884011c8-6587-11e7-8745-1dc353cf7249.png)

**Comment by jenlampton** [comment]
I like the looks of this, it's easy to read, and clear that this post is unpublished. Plus, I think it will work on teasers as well. I'd love to get some more feedback.

**Comment by wesruv** [comment]
Could this be more of a "Mark" like: http://backdrop-ops.github.io/styleguide/#mark_new ?

Could maintain the text, but I think it'd be better with the styles of the mark, the bg/text color of http://backdrop-ops.github.io/styleguide/#error-message

**Comment by wesruv** [comment]
Could be crammed into the `$title_suffix`?

**Comment by jenlampton** [comment]
The `$title_suffix` doesn't always have content, so we couldn't put it there with a psuedo-element.  But if we want to add the content type name into the text, a psuedo-element wouldn't work for that anyway, since we'd need to pull that human-readable name from config. 

We could add something in via preprocess for unpublished nodes, but that would be a Basis-only markup change, where all the other themes depend on only the existing CSS class. I don't have a problem with this approach personally, but wanted to point it out.

**Comment by olafgrabienski** [comment]
> if we want to add the content type name into the 'Unpublished X' text, a psuedo-element wouldn't work for that anyway (...) We could add something in via preprocess for unpublished nodes, but that would be a Basis-only markup change, where all the other themes depend on only the existing CSS class.

Im my opinion, preprocess for _Basis_ only is fine as the addition of the content type information is an enhancement but it's not necessary.

(And a question: can the preprocess addition be copied to other themes?)

**Comment by jenlampton** [comment]
> can the preprocess addition be copied to other themes by a themer / site builder

The code would need to be copied into the new theme's `template.php` file, but I'd guess copying code is definitely something a theme developer could do, and probably something a site builder could do as well, with the proper instructions.

**Comment by jenlampton** [comment]
Here's what it looks like when it's crammed into title_suffix:
<img width="1232" alt="screen shot 2017-08-03 at 12 10 09 pm" src="https://user-images.githubusercontent.com/397895/28939031-caf9a5e0-7844-11e7-8c9d-ea4a22df0e41.png">

**Comment by jenlampton** [comment]
Here's a version with a red border, and I was playing at more human-friendly text too.
<img width="1194" alt="screen shot 2017-08-03 at 12 22 56 pm" src="https://user-images.githubusercontent.com/397895/28939543-8871bf80-7846-11e7-8f0a-e18440c33433.png">

**Comment by jenlampton** [comment]
Here's a bright red background with white text:
<img width="1196" alt="screen shot 2017-08-03 at 12 23 51 pm" src="https://user-images.githubusercontent.com/397895/28939604-aacd42ac-7846-11e7-82f2-38bbfe2c69dc.png">

**Comment by serundeputy** [comment]
Code looks good.
Works good.

**Comment by opi** [comment]
I made a quick code review, and ask for little changes https://github.com/backdrop/backdrop/pull/1936#pullrequestreview-54299450

**Comment by jenlampton** [comment]
Thanks @opi!  I've changed the tags on this issue to NW until we can address your feedback :)

**Comment by jenlampton** [comment]
Pushed a new PR with the feedback addressed : https://github.com/backdrop/backdrop/pull/1936

**Comment by opi** [comment]
approved https://github.com/backdrop/backdrop/pull/1936#pullrequestreview-55724429

**Comment by jenlampton** [comment]
Thanks for the review @opi! I checked your permissions and it looks like you didn't have access to adjust tags, so I've added you :) You should now!

**Comment by quicksketch** [comment]
I think this might need to wait until 1.8.0 if we're adding new markup to all nodes.

Review at https://github.com/backdrop/backdrop/pull/1936#pullrequestreview-56697864, this PR has an XSS issue.

**Comment by olafgrabienski** [comment]
Visually good, and as far as I understood, the XSS issue is fixed. Are there still any other issues?

**Comment by jenlampton** [comment]
Nope, I think this is RTBC?

**Comment by olafgrabienski** [comment]
> I think this is RTBC?

Okay, I've marked it as RTBC! (If anyone has objections, feel free to mark as needed.)

**Comment by quicksketch** [comment]
This works great but does have one more issue: it breaks contextual links on nodes. If the unpublished indicator is present, the contextual links are wiped out. Left a comment on the PR: https://github.com/backdrop/backdrop/pull/1936#pullrequestreview-59080326.

XSS issue is indeed fixed. Other than this issue, PR looks good to go.

**Comment by jenlampton** [comment]
I've pushed a new PR that fixes the contextual link issue. Ready for another review.

**Comment by olafgrabienski** [comment]
I'm not shure *which* contextual links were broken. Tested a view with unpublished node teasers, and these contextual links are present now - see screenshot:

![screen-backdrop-pr-1936-contextual-links](https://user-images.githubusercontent.com/14188307/30034359-1e2181c8-91a1-11e7-8c2b-de689a4686ac.png)

**Comment by opi** [comment]
@olafgrabienski issue with contextual links have been fixed already, it was https://github.com/backdrop/backdrop/pull/1936#pullrequestreview-59080326 (click on "show outdated").
I guess if you have not seen any bug, then it's RTBC ;)

**Comment by quicksketch** [comment]
Great thanks @olafgrabienski, @opi, @jenlampton, @wesruv, @laryn, et al! Merged backdrop/backdrop#1936 into 1.x for 1.8.0.
